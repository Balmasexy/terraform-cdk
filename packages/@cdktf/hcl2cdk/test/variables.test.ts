/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { testCase, Synth } from "./helpers/convert";

describe("variables", () => {
  testCase.test(
    "simple variable",
    `
    variable "image_id" {
        type = string
      }
    
    output "cidr_out" {
        value = var.image_id
    }
    `,
    [],
    Synth.never,
    {
      resources: [],
    }
  );

  testCase.test(
    "variable with type",
    `
    variable "availability_zone_names" {
        type    = list(string)
        default = ["us-west-1a"]
    }
    
    output "cidr_out" {
        value = var.availability_zone_names
    }
    `,
    [],
    Synth.never,
    {
      resources: [],
    }
  );

  testCase.test(
    "variable with complex type",
    `
    variable "docker_ports" {
        type = list(object({
          internal = number
          external = number
          protocol = string
        }))
        default = [
          {
            internal = 8300
            external = 8300
            protocol = "tcp"
          }
        ]
        description = "List of port mappings for the Docker container. Each mapping is exposed as a separate ContainerPort in the container definition."
        sensitive  = true
        nullable   = true
    }

    
    output "cidr_out" {
        value = var.docker_ports
    }
    `,
    [],
    Synth.never,
    {
      resources: [],
    }
  );

  testCase.test(
    "variable with validation",
    `
    variable "image_id" {
        type        = string
        description = "The id of the machine image (AMI) to use for the server."
      
        validation {
            condition     = length(var.image_id) > 4 && substr(var.image_id, 0, 4) == "ami-"
            error_message = "The image_id value must be a valid AMI id, starting with \\"ami-\\"."
          }
      }
      
    
    output "cidr_out" {
        value = var.image_id
    }
    `,
    [],
    Synth.never,
    {
      resources: [],
    }
  );
});
