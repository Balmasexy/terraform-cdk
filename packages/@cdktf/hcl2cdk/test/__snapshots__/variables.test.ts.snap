// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variables simple variable snapshot 1`] = `
"import * as cdktf from \\"cdktf\\";
/*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
You can read more about this at https://cdk.tf/variables*/
const imageId = new cdktf.TerraformVariable(this, \\"image_id\\", {
  type: cdktf.VariableType.STRING
});
new cdktf.TerraformOutput(this, \\"cidr_out\\", {
  value: imageId.value,
});
"
`;

exports[`variables variable with complex type snapshot 1`] = `
"import * as cdktf from \\"cdktf\\";
/*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
You can read more about this at https://cdk.tf/variables*/
const dockerPorts = new cdktf.TerraformVariable(this, \\"docker_ports\\", {
  default: [
    {
      external: 8300,
      internal: 8300,
      protocol: \\"tcp\\",
    },
  ],
  description:
    \\"List of port mappings for the Docker container. Each mapping is exposed as a separate ContainerPort in the container definition.\\",
  nullable: true,
  sensitive: true,
  type: cdktf.VariableType.list(cdktf.VariableType.object({
          internal: cdktf.VariableType.NUMBER,
          external: cdktf.VariableType.NUMBER,
          protocol: cdktf.VariableType.STRING
        }))
});
new cdktf.TerraformOutput(this, \\"cidr_out\\", {
  value: dockerPorts.value,
});
"
`;

exports[`variables variable with type snapshot 1`] = `
"import * as cdktf from \\"cdktf\\";
/*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
You can read more about this at https://cdk.tf/variables*/
const availabilityZoneNames = new cdktf.TerraformVariable(
  this,
  \\"availability_zone_names\\",
  {
    default: [\\"us-west-1a\\"],
    type: cdktf.VariableType.list(cdktf.VariableType.STRING)
  }
);
new cdktf.TerraformOutput(this, \\"cidr_out\\", {
  value: availabilityZoneNames.value,
});
"
`;

exports[`variables variable with validation snapshot 1`] = `
"import * as cdktf from \\"cdktf\\";
"
`;
